internal/infrastructure/postgre/balance.go

```diff
@@ -0,0 +1,269 @@
+                                withdrawal, 
+                                order_number, 
+                                processed_at, 
+                                uuid
+                                ) VALUES ($1, $2, $3, $4, $5, $6)`,
+		operation.UserUUID,
+		operation.Accrual,
+		operation.Withdrawal,
+		operation.OrderNumber,
+		operation.ProcessedAt,
+		operation.UUID)
+	if err != nil {
+		log.Error("Failed to create balance operation", log.ErrorField(err))
+		r.rollback(tx, ctx)
+		return fmt.Errorf("%s: %w", op, err)
+	}
+	err = tx.Commit(ctx)
+	if err != nil {
+		log.Error("Failed to commit transaction", log.ErrorField(err))
+		r.rollback(tx, ctx)
+		return fmt.Errorf("%s: %w", op, err)
+	}
+	return nil
+}
+
+func (r *BalanceRepository) Accrue(ctx context.Context, operation entity.BalanceOperation) error {
+	const op = "infrastructure.postgre.BalanceRepository.Accrue"
+	log := r.log.With(
+		r.log.StringField("op", op),
+		r.log.StringField("request_id", contexter.GetRequestID(ctx)),
+		r.log.StringField("user_uuid", operation.UserUUID.String()),
+		r.log.AnyField("order_number", operation.OrderNumber),
+		r.log.AnyField("accrual", operation.Accrual),
+		r.log.AnyField("withdrawal", operation.Withdrawal),
+	)
+
+	tx, err := r.db.BeginTx(ctx, pgx.TxOptions{})
+	if err != nil {
+		log.Error("Failed to begin transaction", log.ErrorField(err))
+		return fmt.Errorf("%s: %w", op, err)
+	}
+
+	var current float64
+	balance := tx.QueryRow(ctx, `SELECT current FROM user_balances WHERE user_uuid = $1 FOR UPDATE`, operation.UserUUID)
+	err = balance.Scan(&current)
+	if err != nil {
+		log.Error("Failed to get current balance", log.ErrorField(err))
+		r.rollback(tx, ctx)
+		return fmt.Errorf("%s: %w", op, err)
+	}
+
+	_, err = tx.Exec(ctx, `UPDATE user_balances SET current = current + $1 WHERE user_uuid = $2`, operation.Accrual, operation.UserUUID)
+	if err != nil {
+		log.Error("Failed to update balance", log.ErrorField(err))
+		r.rollback(tx, ctx)
+		return fmt.Errorf("%s: %w", op, err)
+	}
+	_, err = tx.Exec(ctx, `INSERT INTO balance_operations (
+                                user_uuid, 
+                                accrual, 
+                                withdrawal, 
+                                order_number, 
+                                processed_at, 
+                                uuid
+                                ) VALUES ($1, $2, $3, $4, $5, $6)`,
+		operation.UserUUID,
+		operation.Accrual,
+		operation.Withdrawal,
+		operation.OrderNumber,
+		operation.ProcessedAt,
+		operation.UUID)
+	if err != nil {
+		log.Error("Failed to create balance operation", log.ErrorField(err))
+		r.rollback(tx, ctx)
+		return fmt.Errorf("%s: %w", op, err)
+	}
+	err = tx.Commit(ctx)
+	if err != nil {
+		log.Error("Failed to commit transaction", log.ErrorField(err))
+		r.rollback(tx, ctx)
+		return fmt.Errorf("%s: %w", op, err)
+	}
+	return nil
+}
+
+func (r *BalanceRepository) rollback(tx pgx.Tx, ctx context.Context) {
```

Comment:
```suggestion
func (r *BalanceRepository) rollback(ctx context.Context, tx pgx.Tx) {
```
В Go принято контекст всегда передавать первым параметром.

Reply:
done